File: python/examples/05_svd.py
# %% [markdown]
# # 5. Singular value decomposition
#
# BatchLAS exposes three SVD drivers with different sweet spots:
#
# | Driver | Intended for |
# |---|---|
# | `gesvd` | general-purpose, uses the vendor path where available |
# | `gesvd_blocked` | native blocked path, medium and large matrices |
# | `gesvd_cta` | one work-group per matrix, very small matrices ($n \le 32$) |
#
# All three compute $A = U \, \mathrm{diag}(s) \, V^{H}$.
#
# The underlying two-step pipeline — reduce to bidiagonal form, then run the
# bidiagonal QR iteration — is also available separately as `gebrd_*` and
# `bdsqr`, which the later sections use directly.

# %%
import numpy as np

import batchlas as bl

from _common import batched_general, header, preferred_device, report, section


def bidiagonal_singular_values(d: np.ndarray, e: np.ndarray) -> np.ndarray:
    """Singular values of the upper bidiagonal matrices described by (d, e)."""
    return np.stack([
        np.linalg.svd(np.diag(d[i]) + np.diag(e[i], 1), compute_uv=False)
        for i in range(d.shape[0])
    ])


header("5. Singular value decomposition")

device = preferred_device()
batch = 4

# %% [markdown]
# ## `gesvd` — the general driver

# %%
section("gesvd: A = U diag(s) Vh")

a = batched_general(batch, 24, 24, seed=1)
u, s, vh = bl.gesvd(a, device=device)
reference = np.linalg.svd(a, compute_uv=False)

report("singular value error", float(np.abs(np.sort(s)[:, ::-1] - reference).max()), tol=1e-9)
report("|U diag(s) Vh - A|", float(np.abs(u @ (s[:, :, None] * vh) - a).max()), tol=1e-9)
report("|U^T U - I|", float(np.abs(u.transpose(0, 2, 1) @ u - np.eye(24)).max()), tol=1e-9)

# %% [markdown]
# ### Values only
#
# Skip the vectors when you do not need them — the call returns just `s`.

# %%
section("Values only: skip the vectors when you do not need them")

values = bl.gesvd(a, compute_vectors=False, device=device)
report("shape", values.shape)
report("error", float(np.abs(np.sort(values)[:, ::-1] - reference).max()), tol=1e-9)

# %% [markdown]
# ## The blocked and CTA drivers
#
# `gesvd_cta` is limited to $n \le 32$ and needs a GPU, but for a large batch of
# small matrices it keeps the whole problem resident in one work-group.

# %%
section("gesvd_blocked: native blocked path")

u_b, s_b, vh_b = bl.gesvd_blocked(a, device=device)
report("singular value error", float(np.abs(np.sort(s_b)[:, ::-1] - reference).max()), tol=1e-9)
report("|U diag(s) Vh - A|", float(np.abs(u_b @ (s_b[:, :, None] * vh_b) - a).max()), tol=1e-9)

section("gesvd_cta: one work-group per matrix, for n <= 32")

small = batched_general(batch, 16, 16, seed=2)
small_reference = np.linalg.svd(small, compute_uv=False)
u_c, s_c, vh_c = bl.gesvd_cta(small, device=device)

report("singular value error", float(np.abs(np.sort(s_c)[:, ::-1] - small_reference).max()), tol=1e-9)
report("|U diag(s) Vh - A|", float(np.abs(u_c @ (s_c[:, :, None] * vh_c) - small).max()), tol=1e-9)

# %% [markdown]
# ## Symmetric input
#
# Passing `uplo` tells the driver the matrix is Hermitian so it can take the
# eigensolver fast path. For a symmetric matrix the singular values are the
# absolute eigenvalues.

# %%
section("Symmetric input: declare uplo to take the Hermitian fast path")

sym = batched_general(batch, 16, 16, seed=3)
sym = (sym + sym.transpose(0, 2, 1)) / 2.0

values = bl.gesvd(sym, compute_vectors=False, uplo="lower", device=device)
report(
    "error vs |eigenvalues|",
    float(np.abs(np.sort(values)[:, ::-1] - np.sort(np.abs(np.linalg.eigvalsh(sym)))[:, ::-1]).max()),
    tol=1e-9,
)

# %% [markdown]
# ## `gebrd_*` — reduction to bidiagonal form
#
# All three variants return `(a, d, e, tauq, taup)`: the bidiagonal coefficients
# `d` and `e`, plus the reflectors generating the orthogonal factors $Q$ and $P$.
# The bidiagonal matrix has the same singular values as the input.

# %%
section("gebrd_unblocked / gebrd_cta / gebrd_blocked: reduction to bidiagonal form")

for label, call, matrix in (
    ("unblocked", lambda mat: bl.gebrd_unblocked(mat, device=device), small),
    ("cta      ", lambda mat: bl.gebrd_cta(mat, device=device), small),
    ("blocked  ", lambda mat: bl.gebrd_blocked(mat, block_size=16, device=device), a),
):
    _, d, e, tauq, taup = call(matrix)
    expected = np.linalg.svd(matrix, compute_uv=False)
    got = np.sort(bidiagonal_singular_values(d, e))[:, ::-1]
    report(f"{label} singular value error", float(np.abs(got - expected).max()), tol=1e-9)

# %% [markdown]
# ## `bdsqr` — bidiagonal QR iteration
#
# `gebrd` followed by `bdsqr` is the SVD pipeline spelled out: reduce, then iterate.

# %%
section("bdsqr: bidiagonal QR iteration on the (d, e) coefficients")

reduced, d, e, tauq, taup = bl.gebrd_cta(small, device=device)
singular = bl.bdsqr(d, e, sort_desc=True, device=device)
report("error", float(np.abs(singular - small_reference).max()), tol=1e-9)

# %% [markdown]
# ## `ormbr` — apply the $Q$ or $P$ factor from `gebrd`
#
# This is how the bidiagonal problem's vectors are turned back into vectors of
# the original $A$. Pass the *reduced* matrix returned by `gebrd` — that is where
# the reflectors live.

# %%
section("ormbr: apply the Q or P factor from gebrd")

c = np.tile(np.eye(16), (batch, 1, 1))
q_applied = bl.ormbr(reduced, tauq, c, vect="Q", side="left", trans="n", device=device)
report("|Q^T Q - I|", float(np.abs(q_applied.transpose(0, 2, 1) @ q_applied - np.eye(16)).max()), tol=1e-9)

File: examples/cpp/05_svd.cc
// 5. Singular value decomposition
//
// The gesvd driver and its native variants, plus the pieces they are built
// from: gebrd (reduction to bidiagonal form), bdsqr (bidiagonal QR iteration)
// and ormbr (applying the reduction's reflectors).
//
// A = U * diag(s) * V^H, with s descending.

#include <cstddef>
#include <iostream>

#include <blas/linalg.hh>
#include <util/sycl-vector.hh>

#include "example_utils.hh"

using namespace batchlas;
using namespace examples;

namespace {

constexpr int kBatch = 2;

template <Backend B>
struct Example {
    static void run(Queue& ctx) {
        gesvd_section(ctx);
        values_only_section(ctx);
        variants_section(ctx);
        symmetric_section(ctx);
        gebrd_section(ctx);
        bdsqr_section(ctx);
        ormbr_section(ctx);
    }

    // gesvd — the general driver.
    //
    // A is overwritten. You supply U (m x m), V^H (n x n) and a
    // `singular_values` span of length min(m, n) per batch item — REAL, even
    // when A is complex, hence `Span<float_t<T>>` in the signature.
    // SvdVectors::All asks for the vectors, SvdVectors::None skips them.
    static void gesvd_section(Queue& ctx) {
        section("gesvd - the general driver");

        const int m = 6, n = 4, k = std::min(m, n);
        auto A = Matrix<double>::Random(m, n, false, kBatch, 11);
        auto U = Matrix<double>::Zeros(m, m, kBatch);
        auto Vh = Matrix<double>::Zeros(n, n, kBatch);
        UnifiedVector<double> s(static_cast<size_t>(k) * kBatch);

        UnifiedVector<std::byte> ws(gesvd_buffer_size<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(),
                                                          SvdVectors::All, SvdVectors::All));
        gesvd<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::All, SvdVectors::All, ws.to_span());
        ctx.wait();

        print_values("singular values (descending)", s.to_span(), k);
        print("U shape", std::to_string(U.rows()) + "x" + std::to_string(U.cols()));
        print("V^H shape", std::to_string(Vh.rows()) + "x" + std::to_string(Vh.cols()));
        std::cout << "U, item 0:\n";
        U.view()[0].print(std::cout, 6, 6);
    }

    // Values only
    //
    // SvdVectors::None skips the back-transforms. U and V^H still have to be
    // passed — there is no overload without them — but they are not written.
    static void values_only_section(Queue& ctx) {
        section("Values only");

        const int m = 6, n = 4, k = std::min(m, n);
        auto A = Matrix<double>::Random(m, n, false, kBatch, 11);
        auto U = Matrix<double>::Zeros(m, m, kBatch);
        auto Vh = Matrix<double>::Zeros(n, n, kBatch);
        UnifiedVector<double> s(static_cast<size_t>(k) * kBatch);

        UnifiedVector<std::byte> ws(gesvd_buffer_size<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(),
                                                          SvdVectors::None, SvdVectors::None));
        gesvd<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::None, SvdVectors::None, ws.to_span());
        ctx.wait();
        print_values("same values, no vectors computed", s.to_span(), k);
    }

    // The native variants
    //
    // gesvd dispatches to a vendor library. `gesvd_blocked` and `gesvd_cta` are
    // BatchLAS's own implementations, callable directly when you want a
    // specific one: blocked for medium sizes, CTA for very small square
    // matrices (n <= 32, one work-group per matrix, so GPU only).
    static void variants_section(Queue& ctx) {
        section("The blocked and CTA drivers");

        const int n = 6;
        auto solve = [&](auto&& call, const char* label) {
            auto A = Matrix<double>::Random(n, n, false, kBatch, 21);
            auto U = Matrix<double>::Zeros(n, n, kBatch);
            auto Vh = Matrix<double>::Zeros(n, n, kBatch);
            UnifiedVector<double> s(static_cast<size_t>(n) * kBatch);
            call(A, s, U, Vh);
            ctx.wait();
            print_values(label, s.to_span(), n);
        };

        solve(
            [&](auto& A, auto& s, auto& U, auto& Vh) {
                UnifiedVector<std::byte> ws(gesvd_blocked_buffer_size<B>(ctx, A.view(), s.to_span(), U.view(),
                                                                         Vh.view(), SvdVectors::All, SvdVectors::All));
                gesvd_blocked<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::All, SvdVectors::All,
                                 ws.to_span());
            },
            "gesvd_blocked");

        if constexpr (has_cta_variants<B>) {
            if (supports_cta(ctx)) {
                solve(
                    [&](auto& A, auto& s, auto& U, auto& Vh) {
                        UnifiedVector<std::byte> ws(gesvd_cta_buffer_size<B>(
                            ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::All, SvdVectors::All));
                        gesvd_cta<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::All,
                                     SvdVectors::All, ws.to_span());
                    },
                    "gesvd_cta");
            } else {
                skip("gesvd_cta", "needs a GPU with sub-group width 32");
            }
        } else {
            skip("gesvd_cta", "not instantiated for this backend");
        }
    }

    // Symmetric input
    //
    // The overload taking a `Uplo` says "this matrix is symmetric/Hermitian,
    // read only that triangle". Note the known defect on NETLIB, where the hint
    // is ignored and the whole matrix is read — passing the full symmetric
    // matrix, as here, is safe everywhere.
    static void symmetric_section(Queue& ctx) {
        section("Symmetric input");

        const int n = 5;
        auto A = Matrix<double>::Random(n, n, /*hermitian=*/true, kBatch, 31);
        auto U = Matrix<double>::Zeros(n, n, kBatch);
        auto Vh = Matrix<double>::Zeros(n, n, kBatch);
        UnifiedVector<double> s(static_cast<size_t>(n) * kBatch);

        UnifiedVector<std::byte> ws(gesvd_buffer_size<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(),
                                                          SvdVectors::All, SvdVectors::All, Uplo::Lower));
        gesvd<B>(ctx, A.view(), s.to_span(), U.view(), Vh.view(), SvdVectors::All, SvdVectors::All, Uplo::Lower,
                 ws.to_span());
        ctx.wait();
        print_values("singular values (= |eigenvalues| for a symmetric matrix)", s.to_span(), n);
    }

    // gebrd — reduction to bidiagonal form.
    //
    // The first half of any SVD: A = Q * Bd * P^H with Bd bidiagonal, returned
    // as its diagonal `d` and superdiagonal `e`. Q and P stay in packed
    // reflector form in A plus the scalars tauq/taup, exactly like geqrf.
    //
    // Three implementations with the same contract: unblocked, cta (small,
    // GPU) and blocked (the only one taking a workspace). Note `e` must have
    // length n-1; the blocked path validates this and throws otherwise.
    static void gebrd_section(Queue& ctx) {
        section("gebrd - reduction to bidiagonal form");

        const int n = 8;

        {
            auto A = Matrix<double>::Random(n, n, false, kBatch, 41);
            Vector<double> d(n, kBatch), e(n - 1, kBatch), tauq(n, kBatch), taup(n, kBatch);
            gebrd_unblocked<B>(ctx, A.view(), VectorView<double>(d), VectorView<double>(e), VectorView<double>(tauq),
                               VectorView<double>(taup));
            ctx.wait();
            std::cout << "gebrd_unblocked diagonal:      ";
            d.batch_item(0).print();
            std::cout << "gebrd_unblocked superdiagonal: ";
            e.batch_item(0).print();
        }

        {
            auto A = Matrix<double>::Random(n, n, false, kBatch, 41);
            Vector<double> d(n, kBatch), e(n - 1, kBatch), tauq(n, kBatch), taup(n, kBatch);
            UnifiedVector<std::byte> ws(gebrd_blocked_buffer_size<B>(ctx, A.view(), VectorView<double>(d),
                                                                      VectorView<double>(e), VectorView<double>(tauq),
                                                                      VectorView<double>(taup), 16));
            gebrd_blocked<B>(ctx, A.view(), VectorView<double>(d), VectorView<double>(e), VectorView<double>(tauq),
                             VectorView<double>(taup), ws.to_span(), /*block_size=*/16);
            ctx.wait();
            std::cout << "gebrd_blocked diagonal:        ";
            d.batch_item(0).print();
        }

        if constexpr (has_cta_variants<B>) {
            if (supports_cta(ctx)) {
                auto A = Matrix<double>::Random(n, n, false, kBatch, 41);
                Vector<double> d(n, kBatch), e(n - 1, kBatch), tauq(n, kBatch), taup(n, kBatch);
                gebrd_cta<B>(ctx, A.view(), VectorView<double>(d), VectorView<double>(e), VectorView<double>(tauq),
                             VectorView<double>(taup));
                ctx.wait();
                std::cout << "gebrd_cta diagonal:            ";
                d.batch_item(0).print();
            } else {
                skip("gebrd_cta", "needs a GPU with sub-group width 32");
            }
        } else {
            skip("gebrd_cta", "not instantiated for this backend");
        }
    }

    // bdsqr — bidiagonal QR iteration.
    //
    // The second half: singular values from (d, e). The values-only overload
    // just writes them out; the other also accumulates the rotations into
    // matrices you supply, which is how the full SVD gets its vectors.
    static void bdsqr_section(Queue& ctx) {
        section("bdsqr - bidiagonal QR iteration");

        const int n = 8;
        auto A = Matrix<double>::Random(n, n, false, kBatch, 51);
        Vector<double> d(n, kBatch), e(n - 1, kBatch), tauq(n, kBatch), taup(n, kBatch);
        gebrd_unblocked<B>(ctx, A.view(), VectorView<double>(d), VectorView<double>(e), VectorView<double>(tauq),
                           VectorView<double>(taup));
        ctx.wait();

        UnifiedVector<double> s(static_cast<size_t>(n) * kBatch);
        UnifiedVector<std::byte> ws(bdsqr_buffer_size<double>(ctx, d, e, s.to_span()));
        bdsqr<B>(ctx, VectorView<double>(d), VectorView<double>(e), s.to_span(), ws.to_span(), /*sort_desc=*/true);
        ctx.wait();

        print_values("gebrd then bdsqr: singular values", s.to_span(), n);
    }

    // ormbr — apply the Q or P factor from gebrd.
    //
    // `vect` picks which: 'Q' uses tauq (the left reflectors), 'P' uses taup
    // (the right ones). This is the back-transform that turns the bidiagonal
    // problem's vectors into A's.
    static void ormbr_section(Queue& ctx) {
        section("ormbr - apply the Q or P factor from gebrd");

        const int n = 6;
        auto A = Matrix<double>::Random(n, n, false, kBatch, 61);
        Vector<double> d(n, kBatch), e(n - 1, kBatch), tauq(n, kBatch), taup(n, kBatch);
        gebrd_unblocked<B>(ctx, A.view(), VectorView<double>(d), VectorView<double>(e), VectorView<double>(tauq),
                           VectorView<double>(taup));
        ctx.wait();

        auto C = Matrix<double>::Identity(n, kBatch);
        UnifiedVector<std::byte> ws(ormbr_buffer_size<B>(ctx, A.view(), VectorView<double>(tauq), C.view(), 'Q',
                                                          Side::Left, Transpose::NoTrans));
        ormbr<B>(ctx, A.view(), VectorView<double>(tauq), C.view(), 'Q', Side::Left, Transpose::NoTrans,
                 ws.to_span());
        ctx.wait();

        std::cout << "Q applied to the identity, i.e. Q itself, item 0:\n";
        C.view()[0].print();

        ormbr<B>(ctx, A.view(), VectorView<double>(tauq), C.view(), 'Q', Side::Left, Transpose::Trans, ws.to_span());
        ctx.wait();
        std::cout << "Q^T Q is the identity again:\n";
        C.view()[0].print();
    }
};

}  // namespace

BATCHLAS_EXAMPLE_MAIN("5. Singular value decomposition")
